velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/aggregates/AverageAggregate.h"
+#include "velox/functions/lib/aggregates/AverageAggregateBase.h"
+
+#include "velox/functions/lib/aggregates/SumCountChecker.h"
+
+using namespace facebook::velox::functions::aggregate;
+
+namespace facebook::velox::functions::aggregate::sparksql {
+namespace {
+
+template <typename TInput, typename TAccumulator, typename TResult>
+class AverageAggregate
+    : public AverageAggregateBase<TInput, TAccumulator, TResult> {
+ public:
+  explicit AverageAggregate(TypePtr resultType)
+      : AverageAggregateBase<TInput, TAccumulator, TResult>(resultType) {}
+
+  void addRawInput(
```

Comment:
Is it necessary to change addRawInput? Can we just modify extractAccumulators to turn nulls into (0, 0) structs?

Reply:
Makes sense. Fixed in this way.